### DISCUSSION TITLE
ESP-Tasmota as Marlin sniffer

### DISCUSSION THREAD

**Discussion by marticasamayor** [arendst/Tasmota#11770] [Q&A] [answered]
Hi everybody,

I'm trying to sniffer Marlin's state from my 3D printer to Home Assistant. I know that already exist "ESP3D" and "Octoprint", but: ESP3D don't has MQTT implemented yet; Octoprint runs over a raspberry; And I just want to know the printing state over the MQTT *Implementing decodification by rules or in Home assistant, doesn't matter.*

## HW connections
My 3D printer works with a RAMPS 1.4 and I'm using an ESP8266-01, connected like this *(inverting TX and RX)* [serial UART](https://tasmota.github.io/docs/Peripherals/#serial)
![](https://raw.githubusercontent.com/wiki/luc-github/ESP3D/images/HW/Wires.png)

And connecting to the RAMPS 1.4 board in AUX1:
![](https://raw.githubusercontent.com/wiki/luc-github/ESP3D/images/RAMPS1.4/RAMPS.PNG)

As explained in [ESP3D](https://github.com/luc-github/ESP3D/wiki/Ramps-1.4-Re-ARM) and [Marlin](https://reprap.org/wiki/RAMPS_1.4#BT_Extension) wikis

## FW
I've flashed *tasmota-display.bin* due has serial communication libraries. So, I've run this template `{"NAME":"Marlin","GPIO":[1,3200,1,3232,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}` and communications started.
![image](https://user-images.githubusercontent.com/15909830/115127214-1559f700-9fd5-11eb-9fcd-8ccc7857f252.png)
As you can see I have some problems:
- *Serial buffer overrun*: seems I have to compile my own build (right¿?) and [resize the buffe](https://github.com/arendst/Tasmota/issues/8829#issuecomment-652511009). For example using this [easy way](https://gitpod.io/#https://github.com/benzino77/tasmocompiler), but I do not wich `#Definition` command use.
- *Serial Briedge to MQTT*: I don't found any command in [Serial Briedge](https://tasmota.github.io/docs/Commands/#serial-bridge) as SetOption66: *Set publishing TuyaReceived to MQTT*. For sure it can be done using rules, but exist this option?

Any idea to continue working?

Thanks in advance

**Comment by Jason2866** [comment, marked as answer]
Use standard Tasmota.bin. You do not use a Display so no need for.
Have you read this? https://tasmota.github.io/docs/Commands/#serial-bridge
To activate serial bridge you have to send something!
Example: 
```
Rule1
 on System#Boot do Backlog Baudrate 9600; SerialSend5 0 endon
```
Do not forget to disable serial log in Tasmota

**Reply by marticasamayor** [reply]
Thank you, now it's reading Marlin commands!!!
Used this rule:
`Rule1
 on System#Boot do Backlog Baudrate 115200; SerialSend1 0 endon`
and with this *Marlin* commands I can obtine the desired info:
![image](https://user-images.githubusercontent.com/15909830/115158537-ebfda180-a08e-11eb-960a-2c9201e41112.png)
To abstract this info and send to MQTT I tried this rule:
`Rule2
  on SerialReceived#Data="SD printing byte" DO publish2 stat/%topic%/PERCENTAGE %value% endon` but didn't work. I need work more in this

**Reply by sfromis** [reply]
No need for an extraction rule, the data is *already* being sent via MQTT. Rules basically have no features for string processing or cutting up.

**Reply by marticasamayor** [reply]
Right! Listening in `tele/%topic%/RESULT` I recive the values.
Now the idea is configure a sensor in home assistant from tasmota rule *to make a easy way to publish it in tasmota wiki template*

Until now I tried this, but didn't work:
```
rule2 on system#boot do publish2 homeassistant/sensor/%macaddr%_percentage/config { \"name\":\"Percentage\", \"state_topic\":\"tele/%topic%/SENSOR\", \"value_template\":\"{{ value }}\", \"icon\":\"mdi:percent\", \"unique_id\":\"%macaddr%_percentage\", \"device\":{\"connections\":[[\"mac\",\"%macaddr%\"]]} } endon
```
I'll work in it

**Reply by marticasamayor** [reply]
The idea is to configure the "value_template" filtering the diferents "SerialSend" values, something like this:
```yalm
rule2 on system#boot do publish2 homeassistant/sensor/%macaddr%_percentage/config {"name":"Percentatge3D_time","state_topic":"tele/%topic%/RESULT","value_template":"{{ > {% if "{{ value_json[SerialReceived].split(:)[1] }}", "Print time" %} {{ value_json[SerialReceived].split(:)[2].split(\n)[0] }}{% endif %} }}","icon":"mdi:percent","unique_id":"%macaddr%_percentage3D","device":{"connections":[["mac", "%macaddr%"]]}} endon
```
The split works properly, but the conditional operation didn't, I supose it's a typologi issue

Opened question in Home assistant site: https://community.home-assistant.io/t/value-template-condition/301742

**Reply by marticasamayor** [reply]
Seems that I found a Tasmota limits, haha

I've tried to use this 3 rules, but seems it oversize the flash memory assigned to the rules part:
```
rule1 on system#boot do backlog 
Baudrate 115200;
SerialSend1 0;
publish2 homeassistant/sensor/%macaddr%_bedTemp/config {"name":"Bed temp/setpoint","state_topic":"tele/%topic%/RESULT","value_template":"{{ > {% if value_json["SerialReceived"].split(":")[0].split(" ")[0] == 'ok' %} {{ value_json["SerialReceived"].split(":")[2].split("@")[0] }} {% else %} {{ N/A }} {% endif %} }}","icon":"mdi:tray","unit_of_measurement":"ºC","unique_id":"%macaddr%_bedTemp","device":{"connections":[["mac", "%macaddr%"]]}};
publish2 homeassistant/sensor/%macaddr%_nozzleTemp/config {"name":"Nozzle temp/setpoint","state_topic":"tele/%topic%/RESULT","value_template":"{{ > {% if value_json["SerialReceived"].split(":")[0].split(" ")[0] == 'ok' %} {{ value_json["SerialReceived"].split(":")[1].split("B")[0] }} {% else %}  {{ N/A }} {% endif %} }}","icon":"mdi:printer-3d-nozzle","unit_of_measurement":"ºC","unique_id":"%macaddr%_nozzleTemp","device":{"connections":[["mac", "%macaddr%"]]}};
RuleTimer1 600
endon
```
(*The sensor configuration works in config.yalm of Home Assistant*)
```
rule2 on system#boot do backlog
publish2 homeassistant/sensor/%macaddr%_timeElapset/config {"name":"Printing Time","state_topic":"tele/%topic%/RESULT","value_template":"{{ > {% if value_json["SerialReceived"].split(":")[1] == 'Print time' %} {{ value_json["SerialReceived"].split(":")[2].split("\n")[0] }} {% else %}  {{ N/A }} {% endif %} }}","icon":"mdi:timelapse","unique_id":"%macaddr%_timeElapset","device":{"connections":[["mac", "%macaddr%"]]}};
publish2 homeassistant/sensor/%macaddr%_percentage/config {"name":"Percentage printed","state_topic":"tele/%topic%/RESULT","value_template":"{{ > {% if value_json["SerialReceived"].split(" ")[0] == 'SD' %} {{ value_json["SerialReceived"].split(" ")[3].split("/")[0] / value_json["SerialReceived"].split(" ")[3].split("/")[1] * 100 }} {% elif value_json["SerialReceived"].split("\n")[0] == 'Not SD printing' %} Not printing {% else %} {{ N/A }} {% endif %} }}","icon":"mdi:percent","unique_id":"%macaddr%_percentPrinted","device":{"connections":[["mac", "%macaddr%"]]}}
endon
```
```
rule3 on clock#Timer=1 do backlog 
SerialSend M27;
RuleTimer2 5;
SerialSend M31;
RuleTimer2 5;
SerialSend M105;
RuleTimer2 5;
RuleTimer1 600
endon
```
At turn ON the rule 1 or rule 2, Tasmota colapse and make a reset

My second option, is make the configuration manually in configuration.yalm of Home Assistant and just run the to rules of initialitzation and timers

**Reply by arendst** [reply]
Try latest released version which probably fixes your "colapse".

**Reply by marticasamayor** [reply]
Yes, I was running over 9.4.0 standard tasmota binary. 
As I saw, just there are available 3 rules, what is the limit length for each rule?
I'll try again

**Reply by sfromis** [reply]
Each of the 3 rule set buffers can house a list of rules up to 511 bytes compressed, which usually means near 3 times that raw text. Totaling up, I'd expect about 4K of uncompressed rules.
